command/job_restart.go

```diff
@@ -0,0 +1,849 @@
+	for restartCount, allocID := range restartAllocs {
+		// Block and wait before each iteration if the command is handling an
+		// interrupt signal.
+		<-activeCh
+
+		// Print new batch header every time we restart a multiple of the batch
+		// size which indicate we're starting a new batch.
+		// Skip batch header if batch size is one because it's redundant.
+		if restartCount%c.batchSize == 0 && c.batchSize > 1 {
+			batchNumber := restartCount/c.batchSize + 1
+			remaining := len(restartAllocs) - restartCount
+
+			c.ui.Output(c.Colorize().Color(fmt.Sprintf(
+				"[bold]==> %s: Restarting %s batch of %d allocations[reset]",
+				formatTime(time.Now()),
+				humanize.Ordinal(batchNumber),
+				helper.Min(c.batchSize, remaining),
+			)))
+		}
+
+		// Restart allocation. Wrap the callback function to capture the
+		// allocID loop variable and prevent if from changing inside the
+		// goroutine at each iteration.
+		restarts.Go(func(allocID string) func() error {
+			return func() error {
+				return c.handleAlloc(allocID)
+			}
+		}(allocID))
+
+		// Check if we restarted enough allocations to complete a batch or if
+		// we restarted the last allocation.
+		batchComplete := (restartCount+1)%c.batchSize == 0
+		restartComplete := restartCount+1 == len(restartAllocs)
+		if batchComplete || restartComplete {
+			// Block and wait for restarts to finish and if the command is
+			// handling an interrupt signal.
+			mErr := restarts.Wait()
+			<-activeCh
+
+			// Exit early if an error happens and -fail-on-error is set or if
+			// the error will happen for all restarts.
+			err := mErr.ErrorOrNil()
+			if err != nil && (c.failOnError || c.isErrNotRecoverable(err)) {
+				break
+			}
+
+			// Exit loop before sleeping or asking for user input if we're done
+			// with the last batch.
+			if restartComplete {
+				break
+			}
+
+			// Exit early if -batch-wait=ask and user provided a negative
+			// answer.
+			if c.batchWaitAsk && !c.shouldProceed() {
+				c.ui.Output("\nJob restart canceled.")
+				return 0
+			}
+
+			// Sleep if -batch-wait is set of if -batch-wait=ask and user
+			// provided an interval.
+			if c.batchWait > 0 {
+				c.ui.Output(c.Colorize().Color(fmt.Sprintf(
+					"[bold]==> %s: Waiting %s before restarting the next batch[reset]",
+					formatTime(time.Now()),
+					c.batchWait,
+				)))
+				time.Sleep(c.batchWait)
+			}
+		}
+	}
+
+	c.ui.Output(c.Colorize().Color(fmt.Sprintf(
+		"[bold]==> %s: Finished job restart[reset]",
+		formatTime(time.Now()),
+	)))
+
+	err = restarts.Wait().ErrorOrNil()
+	if err != nil {
+		c.ui.Error(fmt.Sprintf("\nErrors while restarting job:\n%s", strings.TrimSpace(err.Error())))
+		return 1
+	}
+
+	c.ui.Output("\nAll allocations restarted successfully!")
+	return 0
+}
+
+// parseAndValidate parses and validates the arguments passed to the command.
+//
+// This function mutates the command and is not thread-safe so it must be
+// called only once and early in the command lifecycle.
```

Comment:
I :heart: when we have these kinds of warnings around to protect us from our future self eagerly refactoring :grin: